packages/sources/eth-beacon/src/transport/balance.ts

```diff
@@ -0,0 +1,343 @@
+    validatorBatches.forEach((data) => {
+      data.forEach((validator) => {
+        validators.push(validator)
+        balances.push({
+          address: validator.validator.pubkey,
+          balance: validator.balance,
+        })
+      })
+    })
+
+    // Get validators not found on the beacon chain
+    const unfoundValidators = params.addresses.filter(
+      ({ address }) => !balances.find((balance) => balance.address === address),
+    )
+
+    // If searchLimboValidators param set to true, search deposit events for validators not found on the beacon chain
+    // Otherwise, record 0 for the balance of missing validators
+    if (params.searchLimboValidators) {
+      balances.push(...(await this.searchLimboValidators(unfoundValidators)))
+    } else {
+      // Populate balances list with addresses that were filtered out with a 0 balance
+      // Prevents empty array being returned which would ultimately fail at the reduce step
+      // Keep validators list as is to maintain the response received from consensus client
+      unfoundValidators.forEach(({ address }) => {
+        balances.push({
+          address,
+          balance: '0',
+        })
+      })
+    }
+
+    const response = {
+      data: {
+        result: balances,
+      },
+      statusCode: 200,
+      result: null,
+      timestamps: {
+        providerDataRequestedUnixMs: Date.now(),
+        providerDataReceivedUnixMs: Date.now(),
+        providerIndicatedTimeUnixMs: undefined,
+      },
+    }
+    return response
+  }
+
+  getSubscriptionTtlFromConfig(adapterSettings: BaseEndpointTypes['Settings']): number {
+    return adapterSettings.WARMUP_SUBSCRIPTION_TTL
+  }
+
+  private async queryBeaconChain(
+    url: string,
+    addresses: string,
+    statusList?: string,
+  ): Promise<StateResponseSchema> {
+    const requestConfig = {
+      baseURL: this.config.ETH_CONSENSUS_RPC_URL,
+      url,
+      params: { id: addresses, status: statusList || undefined },
+    }
+
+    const res = await this.requester.request<StateResponseSchema>(
+      calculateHttpRequestKey<BalanceTransportTypes>({
+        context: {
+          adapterSettings: this.config,
+          inputParameters,
+          endpointName: this.endpointName,
+        },
+        data: requestConfig.params,
+        transportName: this.name,
+      }),
+      requestConfig,
+    )
+
+    return res.response.data
+  }
+
+  private async searchLimboValidators(unfoundValidators: Address[]): Promise<BalanceResponse[]> {
+    const balances: BalanceResponse[] = []
+    const limboAddressMap: Record<string, Address> = {}
+    // Parse unfound validators into a map for easier access in limbo search
+    unfoundValidators.forEach((validator) => {
+      limboAddressMap[validator.address.toLowerCase()] = validator
+    })
+
+    // Returns map of validators found in limbo with balances in wei
+    const limboBalances = await this.fetchLimboEthBalances(limboAddressMap, this.provider)
```

Comment:
I think we can also remove `this.provider` from here and use it directly in `fetchLimboEthBalances`